fs-sim/src/System/FS/Sim/Stream.hs

```diff
@@ -20,55 +32,137 @@ import           Test.QuickCheck (Gen)
 
--- | Create a 'Stream' based on the given possibly infinite list of @'Maybe'
--- a@s.
-mkStream :: [Maybe a] -> Stream a
-mkStream = Stream
+{-------------------------------------------------------------------------------
+  Running
+-------------------------------------------------------------------------------}
 
 -- | Advance the 'Stream'. Return the @'Maybe' a@ and the remaining 'Stream'.
+--
+-- Returns 'Nothing' by default if the 'Stream' is empty.
 runStream :: Stream a -> (Maybe a, Stream a)
-runStream s@(Stream [])     = (Nothing, s)
-runStream   (Stream (a:as)) = (a, Stream as)
+runStream s@(Stream _    []    ) = (Nothing, s)
+runStream   (Stream info (a:as)) = (a, Stream info as)
+
+{-------------------------------------------------------------------------------
+  Construction
+-------------------------------------------------------------------------------}
+
+-- | Make an empty 'Stream'.
+empty :: Stream a
+empty = Stream Finite []
 
 -- | Make a 'Stream' that always generates the given @a@.
 always :: a -> Stream a
-always a = Stream (repeat (Just a))
+always x = Stream Infinite (repeat (Just x))
 
--- | Make a 'Stream' generator based on a @a@ generator.
---
--- The generator generates a finite stream of 10 elements, where each element
--- has a chance of being either 'Nothing' or an element generated with the
--- given @a@ generator (wrapped in a 'Just').
+-- | Make a 'Stream' that infinitely repeats the given list.
+repeating :: [Maybe a] -> Stream a
+repeating xs = Stream Infinite $ concat (repeat xs)
+
+-- | UNSAFE: Make a 'Stream' that is marked as definitely finite.
 --
--- The first argument is the likelihood (as used by 'QC.frequency') of a
--- 'Just' where 'Nothing' has likelihood 2.
-mkStreamGen :: Int -> Gen a -> Gen (Stream a)
-mkStreamGen justLikelihood genA =
-    mkStream . dropWhileEnd isNothing <$> replicateM 10 mbGenA
-  where
-    mbGenA = QC.frequency
-      [ (2, return Nothing)
-      , (justLikelihood, Just <$> genA)
-      ]
+-- This is unsafe since a user can pass in any list, and evaluating
+-- 'Test.QuickCheck.shrink' or 'show' on the resulting 'Stream' will diverge. It
+-- is the user's responsibility to only pass in a finite list.
+unsafeMkFinite :: [Maybe a] -> Stream a
+unsafeMkFinite = Stream Finite
+
+-- | Make a 'Stream' that is marked as possibly infinite.
+mkInfinite :: [Maybe a] -> Stream a
+mkInfinite = Stream Infinite
+
+{-------------------------------------------------------------------------------
+  Query
+-------------------------------------------------------------------------------}
 
 -- | Return 'True' if the stream is empty.
 --
 -- A stream consisting of only 'Nothing's (even if it is only one) is not
 -- considered to be empty.
 null :: Stream a -> Bool
-null (Stream []) = True
-null _           = False
+null (Stream _ []) = True
+null _             = False
+
+{-------------------------------------------------------------------------------
+  Generation and shrinking
```

Comment:
We should look into whether it makes sense to test these generators and this shrinker.